File: examples/typescript/financial_report/generate_report.ts
import { AccessPassport } from "../../../src/access-control/accessPassport";
import { PineconeVectorDB } from "../../../src/data-store/vector-DBs/pineconeVectorDB";
import { InMemoryDocumentMetadataDB } from "../../../src/document/metadata/inMemoryDocumentMetadataDB";
import { OpenAIChatModel } from "../../../src/generator/completion-models/openai/openAIChatModel";
import { CSVRetriever } from "../../../src/retrieval/csvRetriever";
import { VectorDBDocumentRetriever } from "../../../src/retrieval/vector-DBs/vectorDBDocumentRetriever";
import { OpenAIEmbeddings } from "../../../src/transformation/embeddings/openAIEmbeddings";
import { AdvisorIdentity } from "./access_control/advisorIdentity";
import {
  FinancialReportDocumentRetriever,
  PortfolioData,
} from "./financialReportDocumentRetriever";
import dotenv from "dotenv";
import { ResourceAccessPolicy } from "../../../src/access-control/resourceAccessPolicy";
import { AlwaysAllowAccessPolicy } from "../../../src/access-control/policies/alwaysAllowAccessPolicy";
import fs from "fs/promises";
import { FinancialReportGenerator } from "./financialReportGenerator";

dotenv.config();

async function main() {
  // Load the metadataDB persisted from ingest_data script
  const metadataDB = await InMemoryDocumentMetadataDB.fromJSONFile(
    "examples/typescript/financial_report/metadataDB.json",
    (key, value) => {
      if (key === "accessPolicies") {
        // deserialize access policies to their instances
        return (value as ResourceAccessPolicy[]).map(
          (policy: ResourceAccessPolicy) => {
            if (policy.policy === "AlwaysAllowAccessPolicy") {
              return new AlwaysAllowAccessPolicy();
            }
            return policy;
            // TODO: Handle other policies for demo. Could also try to dynamically import from
            // policies dir based on name, or have static mapping of policy => class constructor
          }
        );
      }
      return value;
    }
  );

  const vectorDB = await new PineconeVectorDB({
    indexName: "test-financial-report",
    // TODO: Make this dynamic via script param
    namespace: "ea4bcf44-e0f3-46ff-bf66-5b1f9e7502df",
    embeddings: new OpenAIEmbeddings(),
    metadataDB,
  });

  const documentRetriever = new VectorDBDocumentRetriever({
    vectorDB,
    metadataDB,
  });

  // TODO: Make this dynamic via script param
  const portfolioRetriever = new CSVRetriever<PortfolioData>(
    "examples/example_data/financial_report/portfolios/client_a_portfolio.csv"
  );

  const accessPassport = new AccessPassport();
  // const identity = new AdvisorIdentity("client_a"); // TODO: Make this dynamic via script param
  // accessPassport.register(identity);

  const retriever = new FinancialReportDocumentRetriever({
    documentRetriever,
    portfolioRetriever,
    metadataDB,
  });

  const generator = new FinancialReportGenerator();

  const report = await generator.run({
    prompt: "Recovery from the COVIDE-19 pandemic",
    accessPassport, // not necessary in this case, but include for example
    retriever,
  });

  await fs.writeFile("examples/typescript/financial_report/report.txt", report);
}

main();

export {};

File: examples/python/financial_report/generate_report.py
import asyncio
import uuid

from semantic_retrieval.ingestion.data_sources.fs.file_system import FileSystem
from semantic_retrieval.document.metadata.in_memory_document_metadata_db import (
    InMemoryDocumentMetadataDB,
)

from semantic_retrieval.data_store.vector_dbs.pinecone_vector_db import PineconeVectorDB

from semantic_retrieval.transformation.embeddings.openai_embeddings import OpenAIEmbeddings


from semantic_retrieval.retrieval.document_retriever import DocumentRetriever

from semantic_retrieval.generator.retrieval_augmented_generation.vector_db_rag_completion_generator import (
    VectorDBRAGCompletionGenerator,
)

from semantic_retrieval.access_control.access_passport import AccessPassport

from semantic_retrieval.access_control.access_identity import AccessIdentity

from semantic_retrieval.prompts.prompt_template import PromptTemplate


class OpenAIChatModel:
    # TODO:
    pass

    async def run(self, args):
        return "TODO"


class FinancialReportDocumentRetriever(DocumentRetriever):
    def __init__(self, vector_db, metadata_db):
        self.vector_db = vector_db
        self.metadata_db = metadata_db


class FinancialReportGenerator(VectorDBRAGCompletionGenerator):
    def __init__(self, chat_model):
        self.chat_model = chat_model
        super().__init__(chat_model)


async def main():
    # TODO: This needs to be a metadata DB implementation that persists/loads from disk
    metadataDB = InMemoryDocumentMetadataDB()

    vector_db = PineconeVectorDB(
        {
            "indexName": "test-financial-report-py",
            "namespace": "GET NAMESPACE FROM ingest_data RUN",
            "embeddings": OpenAIEmbeddings(),
            "metadataDB": metadataDB,
        }
    )

    retriever = FinancialReportDocumentRetriever(vector_db=vector_db, metadata_db=metadataDB)
    generator = FinancialReportGenerator(OpenAIChatModel())

    accessPassport = AccessPassport()

    access_identity = AccessIdentity("id")
    accessPassport.register(access_identity)

    data = "TODO"

    prompt = PromptTemplate(
        f"Use the following data to construct a financial report matching the following format ... {data}"
    )

    res = await generator.run(
        {
            "accessPassport": accessPassport,
            "prompt": prompt,
            "retriever": retriever,
        }
    )

    print("Report:\n")
    print(res)


if __name__ == "__main__":
    asyncio.run(main())
